File: exercices/solutions/exchange.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char* argv[])
{
    int rank, size;
    float A[2], B[2];
    MPI_Status status;

    MPI_Init( &argc, &argv );
    MPI_Comm_rank( MPI_COMM_WORLD, &rank );
    MPI_Comm_size( MPI_COMM_WORLD, &size );
    if (size < 2) { MPI_Finalize(); return -1; }

    if (rank == 0) {
        A[0] = 1.0f;
        A[1] = 2.0f;

        // Send and Receive
        MPI_Send( &A, 2, MPI_FLOAT, 1, 10, MPI_COMM_WORLD );
        MPI_Recv( &B, 2, MPI_FLOAT, 1, 20, MPI_COMM_WORLD, &status );
        printf("Process %d sent     A=[%f, %f]\n", rank, A[0], A[1]);
        printf("Process %d received B=[%f, %f]\n", rank, B[0], B[1]);
    }

    if (rank == 1) {
        B[0] = 3.0f;
        B[1] = 4.0f;

        // Send and Receive
        MPI_Recv( &A, 2, MPI_FLOAT, 0, 10, MPI_COMM_WORLD, &status );
        MPI_Send( &B, 2, MPI_FLOAT, 0, 20, MPI_COMM_WORLD );
        printf("Process %d sent     B=[%f, %f]\n", rank, B[0], B[1]);
        printf("Process %d received A=[%f, %f]\n", rank, A[0], A[1]);
    }

    MPI_Finalize();
    return 0;
}

File: exercices/solutions/exchange.py
#!/usr/bin/env python

from mpi4py import MPI  # MPI.Init() implicite
import sys


def main():
    """
    Programme principal
    """

    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    if size < 2:
        sys.exit(None)

    if rank == 0:
        A = [1.0, 2.0]

        # Send and Receive
        comm.send(A, dest=1, tag=10)
        B = comm.recv(source=1, tag=20)
        print(f'Process {rank} sent     A=[{A[0]}, {A[1]}]')
        print(f'Process {rank} received B=[{B[0]}, {B[1]}]')

    if rank == 1:
        B = [3.0, 4.0]

        # Send and Receive
        A = comm.recv(source=0, tag=10)
        comm.send(B, dest=0, tag=20)
        print(f'Process {rank} sent     B=[{B[0]}, {B[1]}]')
        print(f'Process {rank} received A=[{A[0]}, {A[1]}]')

    MPI.Finalize()


if __name__ == '__main__':
    main()
